### PR TITLE
chore: Migrate enb_ue_s1ap_id_ue_context_htbl in mme_app to protobuf map

### PR DESCRIPTION
**Pull request by pruthvihebbani**
Migrate enb_ue_s1ap_id_ue_context_htbl in mme_app to protobuf map

## Summary

Migrated enb_ue_s1ap_id_ue_context_htbl in mme_app to protobuf map

## Test Plan

s1ap integ_tests
bazel test //lte/gateway/c/core/...
bazel build --config=production //lte/gateway/c/core:agw_of
bazel test //lte/gateway/c/core/oai/test/amf:amf_stateless_test
export FEATURES=mme_oai; make build_oai

### CODE REVIEW

**Review by ajahl — APPROVED** [review]
Can you rebase, please.

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 1]
```diff
@@ -304,16 +304,14 @@ void mme_app_state_free_ue_context(void** ue_context_node) {
...
     mme_ue_context_t* const mme_ue_context_p, const enb_s1ap_id_key_t enb_key) {
-  hashtable_rc_t h_rc = HASH_TABLE_OK;
-  uint64_t mme_ue_s1ap_id64 = 0;
+  uint32_t mme_ue_s1ap_id = INVALID_MME_UE_S1AP_ID;
 
-  hashtable_uint64_ts_get(mme_ue_context_p->enb_ue_s1ap_id_ue_context_htbl,
-                          (const hash_key_t)enb_key, &mme_ue_s1ap_id64);
-  if (HASH_TABLE_OK == h_rc) {
+  mme_ue_context_p->enb_ue_s1ap_key2mme_ueid_map.get(enb_key, &mme_ue_s1ap_id);
+  if (INVALID_MME_UE_S1AP_ID != mme_ue_s1ap_id) {
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: Put the constant value on RHS

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 2]
```diff
@@ -427,30 +425,36 @@ void mme_ue_context_update_coll_keys(
...
+            "enb_ue_s1ap_ue_id " ENB_UE_S1AP_ID_FMT
+            " mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT,
+            ue_context_p->enb_ue_s1ap_id, ue_context_p->mme_ue_s1ap_id);
+      } else {
+        ue_context_p->enb_s1ap_id_key = enb_s1ap_id_key;
+      }
+    } else {
+      OAILOG_ERROR_UE(
+          LOG_MME_APP, imsi,
+          "Failed to update enb_s1ap_id_key to enb_ue_s1ap_key2mme_ueid_map "
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: Failed to insert enb_s1ap_id_key to enb_ue_s1ap_key2mme_ueid_map

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 3]
```diff
@@ -792,28 +784,26 @@ void mme_remove_ue_context(mme_ue_context_t* const mme_ue_context_p,
...
-      OAILOG_ERROR_UE(LOG_MME_APP, ue_context_p->emm_context._imsi64,
-                      "Failed to remove s11_teid2mme_ueid_map for "
-                      " enb_ue_s1ap_id " ENB_UE_S1AP_ID_FMT
-                      " mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT
-                      ", MME S11 TEID  " TEID_FMT,
-                      ue_context_p->enb_ue_s1ap_id,
-                      ue_context_p->mme_ue_s1ap_id, ue_context_p->mme_teid_s11);
+      OAILOG_ERROR_UE(
+          LOG_MME_APP, ue_context_p->emm_context._imsi64,
+          "Failed to remove s11_teid2mme_ueid_map for s11_teid2mme_ueid_map"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: Failed to remove mme_teid_s11 from s11_teid2mme_ueid_map

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 4]
```diff
@@ -851,14 +841,12 @@ void mme_ue_context_update_ue_sig_connection_state(
...
   if (ue_context_p->ecm_state == ECM_CONNECTED && new_ecm_state == ECM_IDLE) {
-    hash_rc = hashtable_uint64_ts_remove(
-        mme_ue_context_p->enb_ue_s1ap_id_ue_context_htbl,
-        (const hash_key_t)ue_context_p->enb_s1ap_id_key);
-    if (HASH_TABLE_OK != hash_rc) {
+    if ((mme_ue_context_p->enb_ue_s1ap_key2mme_ueid_map.remove(
+            ue_context_p->enb_s1ap_id_key)) != magma::PROTO_MAP_OK) {
       OAILOG_WARNING_UE(LOG_MME_APP, ue_context_p->emm_context._imsi64,
-                        "UE context enb_ue_s1ap_ue_id_key %ld "
+                        "Failed to remove enb_ue_s1ap_ue_id_key %u "
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp:856]: Correct the format specifier for "enb_ue_s1ap_ue_id_key".Some times it's used "%u" and at other times it's used as "%ld"

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp [hunk 5]
```diff
@@ -749,10 +748,9 @@ void MmeNasStateConverter::proto_to_state(const oai::MmeNasState& state_proto,
   OAILOG_INFO(LOG_MME_APP, "Hashtable TEID 11 => MME UE ID");
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp]: Correct the log message

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 6]
```diff
@@ -304,16 +304,14 @@ void mme_app_state_free_ue_context(void** ue_context_node) {
 //------------------------------------------------------------------------------
 ue_mm_context_t* mme_ue_context_exists_enb_ue_s1ap_id(
     mme_ue_context_t* const mme_ue_context_p, const enb_s1ap_id_key_t enb_key) {
-  hashtable_rc_t h_rc = HASH_TABLE_OK;
-  uint64_t mme_ue_s1ap_id64 = 0;
+  uint32_t mme_ue_s1ap_id = INVALID_MME_UE_S1AP_ID;
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp:307]: Using data type as mme_ue_s1ap_id_t, doesn't need conversion at line number 312

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 7]
```diff
@@ -427,30 +425,36 @@ void mme_ue_context_update_coll_keys(
...
-      OAILOG_ERROR_UE(LOG_MME_APP, imsi,
-                      "Error could not update this ue context %p "
-                      "enb_ue_s1ap_ue_id " ENB_UE_S1AP_ID_FMT
-                      " mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT " %s\n",
-                      ue_context_p, ue_context_p->enb_ue_s1ap_id,
-                      ue_context_p->mme_ue_s1ap_id,
-                      hashtable_rc_code2string(h_rc));
+    mme_ue_context_p->enb_ue_s1ap_key2mme_ueid_map.remove(
+        ue_context_p->enb_s1ap_id_key);
+    if (INVALID_MME_UE_S1AP_ID != mme_ue_s1ap_id) {
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp]: Pu the constants on RHS

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 8]
```diff
@@ -615,30 +619,18 @@ status_code_e mme_insert_ue_context(
...
-
-  if (HASH_TABLE_OK != h_rc) {
-    OAILOG_WARNING(LOG_MME_APP,
-                   "Error could not register this ue context %p "
-                   "enb_ue_s1ap_id " ENB_UE_S1AP_ID_FMT " ue_id 0x%x\n",
-                   ue_context_p, ue_context_p->enb_ue_s1ap_id,
-                   ue_context_p->mme_ue_s1ap_id);
-    OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNerror);
+  // Insert enb_s1ap_id_key to enb_ue_s1ap_key2mme_ueid_map
+  if (ue_context_p->enb_s1ap_id_key != INVALID_ENB_UE_S1AP_ID_KEY) {
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp:633]: Before inserting, hashtable_uint64_ts_is_key_exists was called to validate the presence of key "ue_context_p->enb_s1ap_id_key" in hash list. So similarly we are not checking the presence of this key in map.  Here are only validating whether "ue_context_p->enb_s1ap_id_key" has valid value or not

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp [hunk 9]
```diff
@@ -851,14 +841,12 @@ void mme_ue_context_update_ue_sig_connection_state(
     OAILOG_FUNC_OUT(LOG_MME_APP);
   }
   if (ue_context_p->ecm_state == ECM_CONNECTED && new_ecm_state == ECM_IDLE) {
-    hash_rc = hashtable_uint64_ts_remove(
-        mme_ue_context_p->enb_ue_s1ap_id_ue_context_htbl,
-        (const hash_key_t)ue_context_p->enb_s1ap_id_key);
-    if (HASH_TABLE_OK != hash_rc) {
+    if ((mme_ue_context_p->enb_ue_s1ap_key2mme_ueid_map.remove(
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_context.cpp:853]: Remove hash_rc

**Review by rsarwad — APPROVED** [review]
LGTM

**Pull request merged**